### PR TITLE
BEP028 - Provenance

### PR DESCRIPTION
**Pull request by satra**
This replaces #439.

ping @cmaumet @remiadon 

This is a branch directly owned by the bids-specification allowing any maintainer to merge PRs to this. 

Link to the HTML render of this BEP: 
https://bids-specification--487.org.readthedocs.build/en/487/03-modality-agnostic-files.html#provenance-of-bids-datasets-files-and-derivatives

### CODE REVIEW

File: src/03-modality-agnostic-files.md [hunk 1]
```diff
@@ -304,3 +304,299 @@ code organization of these scripts at the moment.
...
 external source repositories to maximize long term preservation (which would
 suffer if an external repository would not be available anymore).
+
+## Provenance of BIDS datasets, files, and derivatives
+
+Template:
+
+```Text
+- [Dataset level] prov.jsonld
+- [File level] sub-<label>/[ses-<label>/]sub-<label>[_ses-<label>]_prov.jsonld
```
Comment (satra) [inline src/03-modality-agnostic-files.md:314]:
```suggestion
- [File level] sub-<label>/[ses-<label>/]sub-<label>[_ses-<label>]_<suffix>.prov
```
To reflect @effigies suggestion.

File: src/03-modality-agnostic-files.md [hunk 2]
```diff
@@ -304,3 +304,299 @@ code organization of these scripts at the moment.
...
+ they can be aggregated without the need to apply any inheritance principle. 
+
+iv. The provenance file MAY be used to reflect the _provenance of a dataset, 
+a collection of files or a specific file at any level_of the bids hierarchy. 
+
+v. Provenance information SHOULD be anonymized/de-identified as necessary. 
+
+### Examples of provenance in BIDS. 
+
+1. The raw conversion from DICOM images or other instrument native formats
```
Comment (satra) [inline src/03-modality-agnostic-files.md:357]:
```suggestion
 1. The raw conversion from DICOM images or other instrument native formats 
```

File: src/03-modality-agnostic-files.md [hunk 3]
```diff
@@ -304,3 +304,299 @@ code organization of these scripts at the moment.
...
+      "derivedFrom": ["sourcedata/12345-1.dcm", "sourcedata/12345-2.dcm"],
+      "generatedBy": {"started": 2019-01-10T10:00:00"
+                      "associatedWith": {"type": "softwareAgent",
+                                         "name": "dcm2niix",
+                                         "version": "2.0.0"},
+                      "commandLine": "dcm2niix ..."
+                      }
+    }
+   ```
+2. In BIDS derivatives, the consideration of outputs requires knowledge of
```
Comment (satra) [inline src/03-modality-agnostic-files.md:376]:
```suggestion
   ```

 1. In BIDS derivatives, the consideration of outputs requires knowledge of 
```

File: src/03-modality-agnostic-files.md [hunk 4]
```diff
@@ -304,3 +304,299 @@ code organization of these scripts at the moment.
...
+      "generatedBy": {"started": 2019-01-10T10:00:00"
+                      "associatedWith": {"type": "softwareAgent",
+                                         "name": "FreeSurfer",
+                                         "uri": "RRID:SCR_001847",
+                                         "version": "6.0.0"},
+                      "commandLine": "mri_convert ..."
+                      }
+    }
+   ```
+3. Provenance can involve information about people and institutions
```
Comment (satra) [inline src/03-modality-agnostic-files.md:395]:
```suggestion
   ```

 1. Provenance can involve information about people and institutions 
```

File: src/03-modality-agnostic-files.md [hunk 5]
```diff
@@ -304,3 +304,299 @@ code organization of these scripts at the moment.
...
+        },
+      "wasAttributedTo": { 
+         "type": "Person",
+         "name": "Prof. Smith",
+         "uri": "ORCID:0123",
+         "role": "PI"}
+       }
+    }
+   ```
+4. Provenance records can highlight reuse of datasets while providing
```
Comment (satra) [inline src/03-modality-agnostic-files.md:418]:
```suggestion
   ```

 1. Provenance records can highlight reuse of datasets while providing 
```

File: src/03-modality-agnostic-files.md [hunk 6]
```diff
@@ -304,3 +304,293 @@ code organization of these scripts at the moment.
...
+2. See the notes and action items below regarding this conversion run.
+3. Check the rendered output (headings, lists, code blocks, tables) for proper
+   formatting and use a linkchecker before you publish this page.
+
+Conversion notes:
+
+* Docs to Markdown version 1.0β20
+* Tue Mar 24 2020 09:07:42 GMT-0700 (PDT)
+* Source doc: BIDS Extension Proposal XX (BEP0XX): Provenance
+----->
```
Comment (sappelhoff) [inline src/03-modality-agnostic-files.md:596]:
```suggestion
```

File: src/03-modality-agnostic-files.md (hunk 1 above)
Comment (cmaumet) [reply src/03-modality-agnostic-files.md:314]: We just discussed this at OHBM Brainhack 2021 and agreed on using `_<suffix>.jsonld`

File: src/03-modality-agnostic-files.md [hunk 7]
```diff
@@ -304,3 +304,277 @@ code organization of these scripts at the moment.
...
+### Encoding Provenance In BIDS
+
+i. Provenance information SHOULD be included in a BIDS dataset when possible.
+
+ii. Provenance records MUST use the [PROV model](https://www.w3.org/TR/prov-o/) 
+ontology and SHOULD be augmented by terms curated in the BIDS specification, 
+the [NIDM](http://nidm.nidash.org/) model, and future enhancements to these models.
+
+iii. If provenance records are included, these records of provenance of a dataset 
+or a file MUST be described using a `[&lt;prefix>_]prov.jsonld` file.
```
Comment (guiomar) [inline src/03-modality-agnostic-files.md:346]:
```suggestion
or a file MUST be described using a `[<prefix>_]prov.jsonld` file. 
```

File: src/03-modality-agnostic-files.md [hunk 8]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+      "derivedFrom": ["sourcedata/12345-1.dcm", "sourcedata/12345-2.dcm"],
+      "generatedBy": {"started": 2019-01-10T10:00:00"
+                      "associatedWith": {"type": "softwareAgent",
+                                         "name": "dcm2niix",
+                                         "version": "2.0.0"},
+                      "commandLine": "dcm2niix ..."
+                      }
+    }
+which inputs from the BIDS dataset were used together with what software was 
+run in what environment and with what parameters.
```
Comment (guiomar) [inline src/03-modality-agnostic-files.md:471]: Not sure what these lines at the end of the examples mean?

File: src/03-modality-agnostic-files.md [hunk 9]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+      "derivedFrom": "sub-01/anat/..._T1.nii.gz",
+      "generatedBy": {"started": 2019-01-10T10:00:00"
+                      "associatedWith": {"type": "softwareAgent",
+                                         "name": "FreeSurfer",
+                                         "uri": "RRID:SCR_001847",
+                                         "version": "6.0.0"},
+                      "commandLine": "mri_convert ..."
+                      }
+    }
+involved in a study.
```
Comment (guiomar) [inline src/03-modality-agnostic-files.md:487]:
```suggestion

```
They seem like not finished?

File: src/03-modality-agnostic-files.md [hunk 10]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+        },
+      "wasAttributedTo": { 
+         "type": "Person",
+         "name": "Prof. Smith",
+         "uri": "ORCID:0123",
+         "role": "PI"}
+       }
+    }
+appropriate attribution to the original dataset generators as well as 
+future transformers.
```
Comment (guiomar) [inline src/03-modality-agnostic-files.md:509]: Also here

File: src/03-modality-agnostic-files.md [hunk 11]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+      "wasAttributedTo": { 
+         "type": "Person",
+         "name": "Prof. Smith",
+         "uri": "ORCID:0123",
+         "role": "PI"}
+       }
+    }
+appropriate attribution to the original dataset generators as well as 
+future transformers.  
+5. For datasets and derivatives, provenance can also include details of
```
Comment (guiomar) [inline src/03-modality-agnostic-files.md:510]: There's something unclosed here so all the following text is reading like in the example/comment grey mode. Maybe it's just missing ``` to close the type of text or something similar. I'm not super expert in this formatting :)

Comment (guiomar) [reply src/03-modality-agnostic-files.md:510]: Also the numbering goes from 1 to 5, so maybe something missing in the middle?

**Comment by sappelhoff** [comment]
(NOTE: I'll cross-post this message across several BEP threads)

Hi there, just a quick notification that we have just merged https://github.com/bids-standard/bids-specification/pull/918 and it may be interesting to look at the implications for this BEP.

We are introducing "BIDS URIs", which unify the way we refer to and point to files in BIDS datasets (as opposed to "dataset-relative" or "subject-relative" or "file-relative" links).

If the diff and discussion in the PR is unclear, you can also read the rendered version: https://bids-specification.readthedocs.io/en/latest/02-common-principles.html#bids-uri

Perhaps there are things in the BEP that need adjusting now, but perhaps also not -- in any case it's good to be aware of this new feature!

Let me know if there are any questions, comments, or concerns.

**Comment by Remi-Gau** [comment]
FYI: added a link to the HTML version of the BEP in the top comment of this PR.

**Comment by Remi-Gau** [comment]
maintenance note: added a link to the rendered BEP in the top message of this PR.

**Comment by yarikoptic** [comment]
while listening to @cmaumet giving a presentation at BIDS derivatives meeting decided to add a reference to a related effort in unrelated other sciences: https://academic.oup.com/gji/article/207/2/1003/2583765 "An Adaptable Seismic Data Format" , ASDF  incorporates the W3C PROV standard to record provenance information

File: src/03-modality-agnostic-files.md [hunk 12]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+external source repositories to maximize long term preservation (which would
+suffer if an external repository would not be available anymore).
+
+## Provenance of BIDS datasets, files, and derivatives
+
+Template:
+
+```Text
+- [Dataset level] prov.jsonld
+- [File level] sub-<label>/[ses-<label>/]sub-<label>[_ses-<label>]_<suffix>.prov
```
Comment (yarikoptic) [inline src/03-modality-agnostic-files.md:409]: Thinking about many types of "applications" operating on entire subject/sessions, indeed it makes sense to have it at the subj/session level (which is presented here, not file level) and have possibility for file level too to "complement" that (similarly how done in inheritance principle for other metadata?):

```suggestion
- [Subject[/session] level] sub-<label>/[ses-<label>/]sub-<label>[_ses-<label>]_<suffix>.prov
- [File level] sub-<label>/[ses-<label>/]<modality>/sub-<label>[_ses-<label>][_entities-<values>]_<suffix>.prov
```

I am also not sure why it is `prov.jsonld` on top and `.prov` and the lower?  why not `.prov.jsonld` at the bottom levels too?

File: src/03-modality-agnostic-files.md [hunk 13]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
 There are no limitations or recommendations on the language and/or
 code organization of these scripts at the moment.
 
+<sup>1</sup>Storing actual source files with the data is preferred over links to
+external source repositories to maximize long term preservation (which would
+suffer if an external repository would not be available anymore).
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply src/03-modality-agnostic-files.md:401]: ha -- this is `Code` section, hence my comment. But upon rereading I wonder if intent was to talk about `sourcedata/`?  Anyways -- language should be clarified. And not 100% sure this addition relates to BEP028 really -- seems generic and could be submitted as independent PR

File: src/03-modality-agnostic-files.md (hunk 12 above)
Comment (yarikoptic) [reply src/03-modality-agnostic-files.md:409]: What about pretty much allowing `.jsonld` at any level where appropriate?  E.g. could be for subject session or specific modality processing etc. All of those to be joined into a single graph anyways, right?

File: src/03-modality-agnostic-files.md [hunk 14]
```diff
@@ -396,6 +396,284 @@ any identifiable information with the source code.
...
+ontology and SHOULD be augmented by terms curated in the BIDS specification, 
+the [NIDM](http://nidm.nidash.org/) model, and future enhancements to these models.
+
+iii. If provenance records are included, these records of provenance of a dataset 
+or a file MUST be described using a `[<prefix>_]prov.jsonld` file. 
+Since these [jsonld](https://json-ld.org/) documents are graph objects,
+they can be aggregated without the need to apply any inheritance principle. 
+
+iv. The provenance file MAY be used to reflect the provenance of a dataset, 
+a collection of files or a specific file at any level_of the bids hierarchy.
```
Comment (yarikoptic) [inline src/03-modality-agnostic-files.md:446]:
```suggestion
a collection of files or a specific file at any level of the bids hierarchy.
```

File: src/03-modality-agnostic-files.md (hunk 12 above)
Comment (yarikoptic) [reply src/03-modality-agnostic-files.md:409]: pushed a fix for `.prov`

**Pull request closed without merging**